float32[] nn_input
---
float32[] nn_output